Userland/Libraries/LibGfx/WebPLoader.cpp

```diff
@@ -0,0 +1,310 @@
+    return chunk;
+}
+
+static ErrorOr<void> decode_webp_simple_lossy(WebPLoadingContext& context, Chunk const& vp8_chunk)
+{
+    // https://developers.google.com/speed/webp/docs/riff_container#simple_file_format_lossy
+    // FIXME
+    (void)context;
+    (void)vp8_chunk;
+    return {};
+}
+
+static ErrorOr<void> decode_webp_simple_lossless(WebPLoadingContext& context, Chunk const& vp8l_chunk)
+{
+    // https://developers.google.com/speed/webp/docs/riff_container#simple_file_format_lossless
+    // FIXME
+    (void)context;
+    (void)vp8l_chunk;
+    return {};
+}
+
+static ErrorOr<void> decode_webp_extended(WebPLoadingContext& context, Chunk const& vp8x_chunk, ReadonlyBytes chunks)
+{
+    // https://developers.google.com/speed/webp/docs/riff_container#extended_file_format
+
+    // FIXME: Do something with this.
+    (void)vp8x_chunk;
+
+    // FIXME: This isn't quite to spec, which says
+    // "All chunks SHOULD be placed in the same order as listed above.
+    //  If a chunk appears in the wrong place, the file is invalid, but readers MAY parse the file, ignoring the chunks that are out of order."
+    while (!chunks.is_empty()) {
+        auto chunk = TRY(decode_webp_advance_chunk(context, chunks));
+
+        if (chunk.type == FourCC("ICCP"))
+            context.icc_data = chunk.data;
+
+        // FIXME: Probably want to make this and decode_webp_simple_lossy/lossless call the same function
+        //        instead of calling the _simple functions from the _extended function.
+        if (chunk.type == FourCC("VP8 "))
+            TRY(decode_webp_simple_lossy(context, chunk));
+        if (chunk.type == FourCC("VP8X"))
+            TRY(decode_webp_simple_lossless(context, chunk));
+    }
+
+    context.state = WebPLoadingContext::State::ChunksDecoded;
+    return {};
+}
+
+static ErrorOr<void> decode_webp_chunks(WebPLoadingContext& context)
+{
+    if (context.state >= WebPLoadingContext::State::ChunksDecoded)
+        return {};
+
+    if (context.state < WebPLoadingContext::HeaderDecoded)
+        TRY(decode_webp_header(context));
+
+    ReadonlyBytes chunks = context.data.slice(sizeof(WebPFileHeader));
+    auto first_chunk = TRY(decode_webp_advance_chunk(context, chunks));
+
+    if (first_chunk.type == FourCC("VP8 ")) {
+        context.state = WebPLoadingContext::State::ChunksDecoded;
+        return decode_webp_simple_lossy(context, first_chunk);
+    }
+
+    if (first_chunk.type == FourCC("VP8L")) {
+        context.state = WebPLoadingContext::State::ChunksDecoded;
+        return decode_webp_simple_lossless(context, first_chunk);
+    }
+
+    if (first_chunk.type == FourCC("VP8X"))
+        return decode_webp_extended(context, first_chunk, chunks);
+
+    return Error::from_string_literal("WebPImageDecoderPlugin: Invalid first chunk type");
+}
+
+WebPImageDecoderPlugin::WebPImageDecoderPlugin(ReadonlyBytes data, OwnPtr<WebPLoadingContext> context)
+{
+    m_context = move(context);
```

Comment:
I don't know what it's called, but we prefer this style in ctors:

```suggestion
    : m_context(move(context))
{
```

Reply:
That's called an initializer list. Done. (In some earlier version of this patch that wasn't possible, and then I forgot to update it when it became possible.)